### PR TITLE
Simplify CallFunc

### PR DESCRIPTION
**Pull request by vgvassilev**
This patch reduces the code bloat in TClingCallFunc trying to make the code more readable.

cc: @sudo-panda

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
@vgvassilev can you make that a PR / ready to review when it's passing the tests, please?

File: core/metacling/src/TClingCallFunc.h [hunk 1]
```diff
@@ -28,16 +28,17 @@
...
+#include "TClingMethodInfo.h"
 #include "TInterpreter.h"
-#include <string>
 
 #include "cling/Interpreter/Value.h"
 
 #include <llvm/ADT/SmallVector.h>
 
+#include <functional>
+#include <string>
```
Comment (vgvassilev) [inline core/metacling/src/TClingCallFunc.h:40]: These will go with the next push I do.

**Comment by vgvassilev** [comment]
@phsft-bot build just on mac1015/cxx17

File: core/metacling/src/TClingCallFunc.cxx [hunk 2]
```diff
@@ -246,29 +177,35 @@ namespace {
    static
    long long sv_to_long_long(const cling::Value &val)
    {
-      return sv_to<long long>(val);
+      return val.getAs<long long>();
```
Comment (Axel-Naumann) [inline core/metacling/src/TClingCallFunc.cxx:180]: Without explicit casting (rather than reinterpret cast like here), the value will be garbled on big endian platforms.

File: core/metacling/src/TClingCallFunc.cxx [hunk 3]
```diff
@@ -917,7 +856,7 @@ int TClingCallFunc::get_wrapper_code(std::string &wrapper_name, std::string &wra
       } break;
       }
    }
-   unsigned min_args = GetMinRequiredArguments();
```
Comment (Axel-Naumann) [inline core/metacling/src/TClingCallFunc.cxx:920]: Calculating the `getMinRequiredArguments()` is a costly operation; we saw it bubble up in several perf profiles. That's why we cache it. Is this not an issue anymore?

File: core/metacling/src/TClingCallFunc.cxx [hunk 4]
```diff
@@ -1210,7 +1149,7 @@ tcling_callfunc_ctor_Wrapper_t TClingCallFunc::make_ctor_wrapper(const TClingCla
    if (const TypeDecl *TD = dyn_cast<TypeDecl>(info->GetDecl())) {
       // This is a class, struct, or union member.
       QualType QT(TD->getTypeForDecl(), 0);
-      ROOT::TMetaUtils::GetNormalizedName(class_name, QT, *fInterp, fNormCtxt);
```
Comment (Axel-Naumann) [inline core/metacling/src/TClingCallFunc.cxx:1213]: Some of the types must not be spelled out with all their template arguments - which is registered in `fNormCtxt`. ATLAS has such a case. @pcanal do you remember the details?

**Review by Axel-Naumann — COMMENTED** [review]
Very nice! I'm not sure about two aspects, see comments. Maybe @pcanal has more insight / better memory?

File: core/metacling/src/TClingCallFunc.cxx (hunk 3 above)
Comment (vgvassilev) [reply core/metacling/src/TClingCallFunc.cxx:920]: Fair point, I thought about that when making this change. I think it will mostly be relevant in the `::exec` call. I will revert that change. However, if `::exec` is the hotspot we do whole bunch of operations there which can be cached as well...

File: core/metacling/src/TClingCallFunc.cxx (hunk 4 above)
Comment (vgvassilev) [reply core/metacling/src/TClingCallFunc.cxx:1213]: `fNormCtxt` tells ROOT to do two things: a) partially desugar the types; b) qualify the types so that the wrapper compiles. We do not practially care about a) (if the wrapper will contain `Double32_t` or `double` and the full type qualification happens in `GetTypeAsString` which calls `cling::utils::Transform::GetPartiallyDesugaredType`.

In fact `hist/histdrawv7/test/histhistdrawv7testUnit` has a test which catches such issues - it just uses horribly long typenames along with inheritance via template parameter packs...

Comment (pcanal) [reply core/metacling/src/TClingCallFunc.cxx:1213]: The issue was:

    In CallFunc remove default arg for __shared_ptr.
    
    We need to manipulate the fully qualified name to remove the default template
    to properly handle the __shared_ptr case until the clang type printer properly
    handles template parameter that are enumerator.

File: core/metacling/src/TClingCallFunc.cxx (hunk 2 above)
Comment (vgvassilev) [reply core/metacling/src/TClingCallFunc.cxx:180]: Hm, are you sure about this? Is there a platform where we can check this?

Or are you simply suggesting `return (long long) val.getAs<long long>()`?

File: core/metacling/src/TClingCallFunc.cxx (hunk 4 above)
Comment (vgvassilev) [reply core/metacling/src/TClingCallFunc.cxx:1213]: @pcanal did you mean 291f23eca4b54ea5166c1f97f01a950620032375

**Comment by vgvassilev** [comment]
@smuzaffar can you try this PR with cmssw? Maybe that will give us another data point.

**Comment by smuzaffar** [comment]
@vgvassilev , cmssw tests are now running for https://github.com/cms-sw/root/pull/172

File: core/metacling/src/TClingCallFunc.cxx (hunk 2 above)
Comment (Axel-Naumann) [reply core/metacling/src/TClingCallFunc.cxx:180]: > Hm, are you sure about this?

Yes.

> Is there a platform where we can check this?

Big endian is almost dead; some Power machines can be booted in big endian mode. You can try here: https://www.godbolt.org/z/zzb4Mx5d9

> Or are you simply suggesting `return (long long) val.getAs<long long>()`?

No, not good enough. You need to access the value as it was stored (`getAs<the-type-as-written-into-the-cling-Value>()`) and then cast that.

File: core/metacling/src/TClingCallFunc.cxx (hunk 4 above)
Comment (pcanal) [reply core/metacling/src/TClingCallFunc.cxx:1213]: The 'real' commit is 'a211a41386e46f8467fdd0c2230d4432d521f542'.   https://github.com/root-project/root/commit/291f23eca4b54ea5166c1f97f01a950620032375 is an improvement on the first one (probably because `GetNormalizedName` is more accurate/better tested than `R__DropDefaultArg`.

The question is whether Clang has improved enough that now, the code generated for the function/method of `__shared_ptr` (or maybe any template) with a (default) template parameter that are enumeraton will now compiled.

Comment (vgvassilev) [reply core/metacling/src/TClingCallFunc.cxx:1213]: How we can check that?

Comment (pcanal) [reply core/metacling/src/TClingCallFunc.cxx:1213]: I assume that using some CallFunc on `shared_ptr<enumtype>` or `shared_ptr<std::vector<enumtype>>` and their base class should be 'enough'.  (technically, but impractically, running the same test with a211a41386e46f8467fdd0c2230d4432d521f542^ and seeing it fail).

File: core/metacling/src/TClingCallFunc.cxx (hunk 2 above)
Comment (vgvassilev) [reply core/metacling/src/TClingCallFunc.cxx:180]: I think I understand what you say now. I think here we are save as we do the recast to the concrete type: we have constructs like `case BuiltinType::Bool: vh.u.b = (bool) sv_to_ulong_long(fArgVals[i]);`

Using the link you sent I discovered something that's important for the further refactoring though.

```cpp
    int i = 17;
    char c = *(char*)&i;
    c= (char) i; // === `c = i;` === `c = (unsigned long long) i;`
    return c;
```

It seems that we can re-cast to the longest type as long as the underlying representation is the expected one. So we don't need to enumerate all types but just enumerate integral (or enum), real, etc.

**Comment by smuzaffar** [comment]
cmssw PR tests look good : https://github.com/cms-sw/root/pull/172#issuecomment-1217329810

**Comment by vgvassilev** [comment]
> cmssw PR tests look good : [cms-sw#172 (comment)](https://github.com/cms-sw/root/pull/172#issuecomment-1217329810)

Thanks a lot @smuzaffar!!

File: core/metacling/src/TClingCallFunc.cxx (hunk 4 above)
Comment (vgvassilev) [reply core/metacling/src/TClingCallFunc.cxx:1213]: I could not build that commit on my platform unfortunately. I added a test which I believe captures the use-case and produces a wrapper which compiles fine:
```cpp
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wformat-security"
__attribute__((used)) __attribute__((annotate("__cling__ptrcheck(off)")))
extern "C" void __cf_11(void* obj, int nargs, void** args, void* ret)
{
   ((void (&)(std::shared_ptr<std::vector<E, std::allocator<E> > >))add_sp)(*(std::shared_ptr<std::vector<E, std::allocator<E> > >*)args[0]);
   return;
}
#pragma clang diagnostic pop
```

File: core/metacling/src/TClingCallFunc.cxx (hunk 2 above)
Comment (pcanal) [reply core/metacling/src/TClingCallFunc.cxx:180]: Wouldn't that depend on the endianess?

File: core/metacling/src/TClingCallFunc.cxx (hunk 4 above)
Comment (pcanal) [reply core/metacling/src/TClingCallFunc.cxx:1213]: Fair enough, it looks (so far) that it cover the case referred to.

File: core/metacling/src/TClingCallFunc.cxx (hunk 2 above)
Comment (Axel-Naumann) [reply core/metacling/src/TClingCallFunc.cxx:180]: I see, we don't store integers of different sizes, so endianess is indeed not an issue:
https://github.com/root-project/root/blob/c2732039eb37cf2fa21f227353aae7e16d3bb935/core/metacling/src/TClingCallFunc.cxx#L2325-L2330


Still, `val.getAs<long long>` on a `val` that doesn't store a `long long` (but say a `double`) can give a wrong value. Unless it's guaranteed that the `Value` stores the same type as being read here, I'd be much happier if we were to use `simplisticCastAs` rather than `getAs`.

Comment (vgvassilev) [reply core/metacling/src/TClingCallFunc.cxx:180]: The simplisticCast comes in #11193. Should I bring the commit here.

Comment (vgvassilev) [reply core/metacling/src/TClingCallFunc.cxx]: Added in this PR.

File: core/metacling/src/TClingCallFunc.cxx [hunk 5]
```diff
@@ -1575,268 +1461,83 @@ void TClingCallFunc::exec(void *address, void *ret)
...
-            //       of the enum here.
-            (void) ET;
+         } else if (QT->isReferenceType() || QT->isRecordType()) {
+            // the argument is already a pointer value (points to the same thing
+            // as the reference or pointing to object passed by value.
+            vp_ary.push_back((void *) fArgVals[i].getAs<unsigned long long>());
+         } else if (QT->isPointerType() || QT->isArrayType() || QT->isMemberPointerType()) {
             ValHolder vh;
-            vh.u.i = (int) sv_to_long_long(fArgVals[i]);
+            vh.u.vp = (void *) fArgVals[i].getAs<unsigned long long>();
```
Comment (Axel-Naumann) [inline core/metacling/src/TClingCallFunc.cxx:1528]:
```suggestion
            vh.u.vp = (void *) fArgVals[i].simplisticCastAs<unsigned long long>();
```

File: core/metacling/src/TClingCallFunc.cxx [hunk 6]
```diff
@@ -1575,268 +1461,83 @@ void TClingCallFunc::exec(void *address, void *ret)
...
-            vp_ary.push_back((void *) sv_to_ulong_long(fArgVals[i]));
-         } else if (const EnumType *ET =
-                    dyn_cast<EnumType>(&*QT)) {
-            // Note: We may need to worry about the underlying type
-            //       of the enum here.
-            (void) ET;
+         } else if (QT->isReferenceType() || QT->isRecordType()) {
+            // the argument is already a pointer value (points to the same thing
+            // as the reference or pointing to object passed by value.
+            vp_ary.push_back((void *) fArgVals[i].getAs<unsigned long long>());
```
Comment (Axel-Naumann) [inline core/metacling/src/TClingCallFunc.cxx:1525]:
```suggestion
            vp_ary.push_back((void *) fArgVals[i].simplisticCastAs<unsigned long long>());
```

File: core/metacling/test/CMakeLists.txt [hunk 7]
```diff
@@ -25,4 +25,6 @@ ROOT_ADD_GTEST(
   TClingMethodInfoTests.cxx
 
   LIBRARIES
-  Core)
+  Core
+  Cling
```
Comment (Axel-Naumann) [inline core/metacling/test/CMakeLists.txt:29]: We must not link against `Cling`. Why is this needed?
```suggestion
```

File: core/metacling/test/TClingCallFuncTests.cxx [hunk 8]
```diff
@@ -15,26 +15,41 @@
 // system, feel free to delete them as these tests here don't represent things the user
 // should do in his code.
 
+// A class that creates a CallFunc and deletes it at the end of its scope.
+class CallFuncRAII {
+  ClassInfo_t *fGlobalNamespace = nullptr;
```
Comment (Axel-Naumann) [inline core/metacling/test/TClingCallFuncTests.cxx:20]:
```suggestion
  ClassInfo_t *fScope = nullptr;
```

File: core/metacling/test/TClingCallFuncTests.cxx [hunk 9]
```diff
@@ -15,26 +15,41 @@
...
+  CallFunc_t *fmc = nullptr;
+public:
+  CallFuncRAII(const char *name) {
```
Comment (Axel-Naumann) [inline core/metacling/test/TClingCallFuncTests.cxx:23]:
```suggestion
  CallFuncRAII(const char *scopeName) {
```

File: core/metacling/test/TClingCallFuncTests.cxx [hunk 10]
```diff
@@ -15,26 +15,41 @@
 // system, feel free to delete them as these tests here don't represent things the user
 // should do in his code.
 
+// A class that creates a CallFunc and deletes it at the end of its scope.
```
Comment (Axel-Naumann) [inline core/metacling/test/TClingCallFuncTests.cxx:18]: Nice!

**Review by Axel-Naumann — APPROVED** [review]
Please address the small comments - but this looks good, thank you! We will see with root-bench whether this costs in terms of performance.

File: core/metacling/test/CMakeLists.txt (hunk 7 above)
Comment (vgvassilev) [reply core/metacling/test/CMakeLists.txt:29]: It is not a link-time dependency, it is a target dependency. If we change TClingCallFunc.cxx and we type `make TClingTest` we want libCling.so to be rebuilt...

Comment (pcanal) [reply core/metacling/test/CMakeLists.txt:29]: Wouldn't something like:
```
add_dependencies(TClingTest Cling)
```
be a better fit for your purpose?

Comment (vgvassilev) [reply core/metacling/test/CMakeLists.txt:29]: Yes, good point.

**Comment by vgvassilev** [comment]
@phsft-bot build just on mac11/cxx14
